crates/re_log_types/src/path/entity_path_filter.rs

```diff
@@ -0,0 +1,244 @@
+/// The last rule matching `world/house` is `+ world/**`, so it is included.
+#[derive(Clone, Default)]
+pub struct EntityPathFilter {
+    rules: BTreeMap<EntityPathRule, RuleEffect>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct EntityPathRule {
+    pub path: EntityPath,
+
+    /// If true, ALSO include children and grandchildren of this path.
+    pub recursive: bool,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum RuleEffect {
+    Include,
+    Exclude,
+}
+
+impl EntityPathFilter {
+    /// Example of rules:
+    ///
+    /// ```diff
+    /// + world/**
+    /// - world/roads/**
+    /// + world/roads/main
+    /// ```
+    ///
+    /// Each line is a rule.
+    ///
+    /// The first character should be `+` or `-`. If missing, `+` is assumed.
+    /// The rest of the line is trimmed and treated as an entity path.
+    ///
+    /// Conflicting rules are resolved by the last rule.
+    pub fn parse_forgiving(rules: &str) -> Self {
+        let mut filter = Self::default();
+
+        for line in rules
+            .split('\n')
+            .map(|line| line.trim())
+            .filter(|line| !line.is_empty())
+        {
+            let (effect, path_pattern) = match line.chars().next() {
+                Some('+') => (RuleEffect::Include, &line[1..]),
+                Some('-') => (RuleEffect::Exclude, &line[1..]),
+                _ => (RuleEffect::Include, line),
+            };
+
+            let rule = EntityPathRule::parse_forgiving(path_pattern);
+
+            filter.add_rule(effect, rule);
+        }
+
+        filter
+    }
+
+    pub fn add_rule(&mut self, effect: RuleEffect, rule: EntityPathRule) {
+        self.rules.insert(rule, effect);
+    }
+
+    pub fn formatted(&self) -> String {
+        let mut s = String::new();
+        for (rule, effect) in &self.rules {
+            s.push_str(match effect {
+                RuleEffect::Include => "+ ",
+                RuleEffect::Exclude => "- ",
+            });
+            s.push_str(&rule.path.to_string());
+            if rule.recursive {
+                s.push_str("/**");
+            }
+            s.push('\n');
+        }
+        if s.ends_with('\n') {
+            s.pop();
+        }
+        s
+    }
+
+    pub fn includes(&self, path: &EntityPath) -> bool {
+        let effect = self
+            .most_specific_match(path)
+            .unwrap_or(RuleEffect::Exclude);
+        match effect {
+            RuleEffect::Include => true,
+            RuleEffect::Exclude => false,
+        }
+    }
+
+    /// Find the most specific matching rule and return its effect.
+    /// If no rule matches, return `None`.
+    pub fn most_specific_match(&self, path: &EntityPath) -> Option<RuleEffect> {
+        // We sort the rule by entity path, with recursive coming before non-recursive.
+        // This means the last matching rule is also the most specific one.
+        for (rule, effect) in self.rules.iter().rev() {
```

Comment:
This implementation works for now for simplicity but there are other data-structures/approaches that should be more efficient -- my intuition is this should be some form of `O(log(n))` rather than `O(n)`.

I need to think through it a bit but something close to `rules.lower_bound(path)` is probably pretty close.

Reply:
Yeah, I thought this would be good enough for a while (matching 1-4 rules on 10-100 entities) but I'll add a TODO for future work